File: Leetcode/0623 Add One Row to Tree.java
class Solution {
    public TreeNode addOneRow(TreeNode root, int val, int depth) {
        if (depth == 1) {
            return new TreeNode(val, root, null);
        }
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        int currDepth = 1;
        while(!q.isEmpty()) {
            ++currDepth;
            if (currDepth == depth) {
                for (var node : q) {
                   node.left = new TreeNode(val, node.left, null);
                   node.right = new TreeNode(val, null, node.right); 
                }
                break;
            }
            int size = q.size();
            for (int i = 0;i < size; i++) {
                var node = q.poll();
                if (node.left != null) q.add(node.left);
                if (node.right != null) q.add(node.right);
            }
        }
        return root;
    }
}

File: Leetcode/0623 Add One Row to Tree.ts
function addOneRow(root: TreeNode | null, val: number, depth: number): TreeNode | null {
    if (depth == 1) {
        return new TreeNode(val, root, null);
    }
    let q: TreeNode[] = [root!];
    let currDepth = 1;
    while (q.length > 0) {
        ++currDepth;
        const newQ: TreeNode[] = [];
        for (const node of q) {
            if (currDepth == depth) {
                node.left = new TreeNode(val, node.left, null);
                node.right = new TreeNode(val, null, node.right);
            } else {
                if (node.left != null) newQ.push(node.left);
                if (node.right != null) newQ.push(node.right);
            }
        }
        q = newQ;
    }
    return root;
}
